FirebaseDatabaseSwift/Sources/Codable/DatabaseReference+WriteEncodable.swift

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseDatabase
+
+extension Database.Encoder {
+  public static var defaultEncoder: () -> Database.Encoder = {
+    .init()
+  }
+}
+
+extension DatabaseReference {
+  /// Encodes an instance of `Encodable` and overwrites the encoded data
+  /// to the path referred by this `DatabaseReference`. If no value exists,
+  /// it is created. If a value already exists, it is overwritten.
+  ///
+  /// See `Database.Encoder` for more details about the encoding process.
+  ///
+  /// - Parameters:
+  ///   - value: An instance of `Encodable` to be encoded to a document.
+  ///   - encoder: An encoder instance to use to run the encoding.
+  ///   - completion: A block to execute once the value has been successfully
+  ///                 written to the server. This block will not be called while
+  ///                 the client is offline, though local changes will be visible
+  ///                 immediately.
+  public func setValue<T: Encodable>(from value: T,
+                                     encoder: Database.Encoder = Database.Encoder.defaultEncoder(),
+                                     completion: ((Error?, DatabaseReference) -> Void)? =
```

Comment:
The DatabaseReference is the same reference from the callsite, so it exists even if the call itself failed. This is the pattern on Android and iOS, but it is notably not used on Web. While I would personally not use this callback style in a new product (and we opted not to for Firestore), I think we should follow the honor the precedent here and continue to use the existing callback signature.

Reply:
For a more 'swifty' API, I would prefer a `Result<Void, Error>` completion handler, so if the `DatabaseReference` can be skipped I would vote for that change. My gut feeling about the intention of this task is not to provide APIs in the same style, but rather design them to fit better with expectations of swift users. But this is just my own take. :-)

With a future (pun somewhat intended) Combine-wrapper API, this would probably translate to returning a (discardableResult) `Future<Void, Error>`. I don't know how the different versions might be disambiguated. This could probably be done with three different overloads - one that does not take a completion handler, one that does and one that returns the `Future`.

Reply:
Wouldn't dropping the DatabaseReference make this call inconsistent with the other APIs in this repo? These APIs are already in our customer's hands and we just missed the window to make breaking changes. There is definitely something to be said about following Swift style, but we should probably also look at other APIs in our SDK for precedent.

Reply:
I'd say that the ongoing work in #6940 suggests that the `Result` type might be a good candidate to use here too?

Reply:
We can use `Result` here.